tls/s2n_psk.c

```diff
@@ -153,6 +153,98 @@ S2N_CLEANUP_RESULT s2n_psk_parameters_wipe(struct s2n_psk_parameters *params)
     return S2N_RESULT_OK;
 }
 
+bool s2n_offered_psk_list_has_next(struct s2n_offered_psk_list *psk_list)
+{
+    return psk_list != NULL && s2n_stuffer_data_available(&psk_list->wire_data) > 0;
+}
+
+S2N_RESULT s2n_offered_psk_list_read_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk)
+{
+    ENSURE_REF(psk_list);
+    ENSURE_MUT(psk);
+
+    uint16_t identity_size = 0;
+    GUARD_AS_RESULT(s2n_stuffer_read_uint16(&psk_list->wire_data, &identity_size));
+    ENSURE_GT(identity_size, 0);
+
+    uint8_t *identity_data = NULL;
+    identity_data = s2n_stuffer_raw_read(&psk_list->wire_data, identity_size);
+    ENSURE_REF(identity_data);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+     *# For identities established externally, an obfuscated_ticket_age of 0 SHOULD be
+     *# used, and servers MUST ignore the value.
+     */
+    GUARD_AS_RESULT(s2n_stuffer_skip_read(&psk_list->wire_data, sizeof(uint32_t)));
+
+    GUARD_AS_RESULT(s2n_blob_init(&psk->identity, identity_data, identity_size));
+    return S2N_RESULT_OK;
+}
+
+int s2n_offered_psk_list_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk)
+{
+    notnull_check(psk_list);
+    notnull_check(psk);
+    *psk = (struct s2n_offered_psk){ 0 };
+    ENSURE_POSIX(s2n_offered_psk_list_has_next(psk_list), S2N_ERR_STUFFER_OUT_OF_DATA);
+    ENSURE_POSIX(s2n_result_is_ok(s2n_offered_psk_list_read_next(psk_list, psk)), S2N_ERR_BAD_MESSAGE);
```

Comment:
Does `S2N_ERR_BAD_MESSAGE` aptly represent this error case? We could consider adding an error code for PSK related errors instead.

Reply:
I'd argue S2N_ERR_BAD_MESSAGE is the right error message.

s2n_offered_psk_list_read_next is parsing part of the ClientHello message / the pre_shared_key extension. We've basically postponed parsing we'd usually do during the recv method to whenever the callback actually needs the information. If we did this parsing during recv we'd definitely use S2N_ERR_BAD_MESSAGE, so I'd argue we should still use S2N_ERR_BAD_MESSAGE when the parsing happens later.